contracts/Ethernauts.sol

```diff
@@ -75,23 +114,48 @@ contract Ethernauts is ERC721Enumerable, Ownable {
         mintPrice = newMintPrice;
     }
 
-    function setBaseURI(string memory baseTokenURI_) public onlyOwner {
-        baseTokenURI = baseTokenURI_;
+    function setEarlyMintPrice(uint newEarlyMintPrice) external onlyOwner {
+        earlyMintPrice = newEarlyMintPrice;
+    }
+
+    function setBaseURI(string memory newBaseTokenURI) external onlyOwner {
+        baseTokenURI = newBaseTokenURI;
+    }
+
+    function setSaleState(SaleState newSaleState) external onlyOwner {
+        require(newSaleState != currentSaleState, "Invalid new state");
+
+        currentSaleState = newSaleState;
+    }
+
+    function setCouponSigner(address newCouponSigner) external onlyOwner {
+        couponSigner = newCouponSigner;
     }
 
     function withdraw(address payable beneficiary) external onlyOwner {
         beneficiary.sendValue(address(this).balance);
     }
 
     // -------------------
-    // Internal functions
+    // Private functions
     // -------------------
 
+    function _verifyCoupon(bytes memory signedCoupon) private {
```

Comment:
I think it would be neat to break out the coupon validation and redemption in two separate functions. 

The validation could be in a separate `view` (maybe even `pure`?) function, and it would just return true if the signed coupon is indeed from `couponSigner`.  That way it would let you verify your coupon gas free before redeeming it. Otherwise a coupon copy-pasta mistake would lead to a costly failure in `_verifyCoupon`. What do you think?

Reply:
Implemented! Pls have a look and let me know if this is of your liking :)

Reply:
Looks great 👍